Kinvey.Core/Core/AbstractKinveyClientRequest.cs

```diff
@@ -491,67 +491,80 @@ public async Task<HttpResponseMessage> ExecuteUnparsedAsync()
-							Client.Store.Store(Client.ActiveUser.Id, Client.SSOGroupKey, currentCred);
-
-							// Retry the original request
-							RequestAuth = new KinveyAuthenticator(currentCred.AuthToken);
-							var retryResponse = await ExecuteUnparsedAsync();
-							return retryResponse;
-						}
-						else
-						{
-							//logout the current user
-							Client.ActiveUser.Logout(); // TODO is this a potential deadlock?
-						}
-					}
-					else
-					{
-						Client.ActiveUser.Logout();
-					}
-				}
-				else
-				{
-					Client.ActiveUser.Logout();
-				}
-			}
+                        if (!string.IsNullOrEmpty(refreshToken) && !refreshToken.ToLower().Equals("null"))
+                        {
+                            if (!hasRetried)
+                            {
+                                // Attempting retry - set flag to prevent additional attempts
+                                hasRetried = true;
+
+                                //use the refresh token for a new access token
+                                JObject result = await Client.ActiveUser.UseRefreshToken(refreshToken, redirectUri, micClientId).ExecuteAsync();
+
+                                // log out the current user without removing the user record from the credential store
+                                Client.ActiveUser.LogoutSoft();
+
+                                //login with the access token
+                                Provider provider = new Provider();
+                                provider.kinveyAuth = new MICCredential(result["access_token"].ToString());
+                                User u = await User.LoginAsync(new ThirdPartyIdentity(provider), Client);
+
+                                //store the new refresh token
+                                Credential currentCred = Client.Store.Load(Client.ActiveUser.Id, Client.SSOGroupKey);
+                                currentCred.AccessToken = result["access_token"].ToString();
+                                currentCred.RefreshToken = result.GetValidValue("refresh_token");
+                                currentCred.RedirectUri = redirectUri;
+                                Client.Store.Store(Client.ActiveUser.Id, Client.SSOGroupKey, currentCred);
+
+                                // Retry the original request
+                                RequestAuth = new KinveyAuthenticator(currentCred.AuthToken);
+                                var retryResponse = await ExecuteUnparsedAsync();
+                                return retryResponse;
+                            }
+                            else
+                            {
+                                Client.ActiveUser.Logout();
+                            }
+                        }
+                        else
+                        {
+                            //logout the current user
+                            Client.ActiveUser.Logout(); // TODO is this a potential deadlock?
```

Comment:
I'm wondering in which scenario this could cause a deadlock?

Reply:
@heyzooi 
As for me I can't see a potential deadlock here. This comment was made a long time ago.
Maybe somebody simply forgot to remove it.

Reply:
@eugeninnowise It looks like something that was potentially relevant in an old version of the SDK, but not now. Please delete the comment before merging.

Reply:
@vinaygahlawat It is done!